### PR TITLE
ExpressLRS support

### PR DESCRIPTION
**Pull request by phobos-**
This adds the same functionality as ExpressLRS LUA in DeviationTX:
https://github.com/ExpressLRS/ExpressLRS/blob/develop/src/lua/ELRS.lua

This code works with ExpressLRS v1.0.0-RC1 onwards.
Attached test binaries for Jumper T8SG V2 Plus.
Feedback is very much welcome.

(Note - only esp based ExpressLRS tx modules are supported -> -DUART_INVERTED needs to be commented out in user_defines.txt)
(Note2 - you need hardware CRSF mod in your radio in order for CRSF to work on the correct pin. Tutorial here: https://www.youtube.com/watch?v=Lr_YRGJl5x8)

[deviation-t8sg_v2_plus-v5.0.0-1dc62dc.zip](https://github.com/DeviationTX/deviation/files/6503349/deviation-t8sg_v2_plus-v5.0.0-1dc62dc.zip)
[deviation-t8sg_v2-v5.0.0-1dc62dc.zip](https://github.com/DeviationTX/deviation/files/6503365/deviation-t8sg_v2-v5.0.0-1dc62dc.zip)

### CODE REVIEW

File: src/protocol/crsf_uart.c [hunk 1]
```diff
@@ -29,6 +29,95 @@
...
+  NULL
+};
+
+static u8 convertPktRateToElrs(u8 rfFreq, u8 rate)
+{
+  switch (rate) {
+    case 0: if (rfFreq == 6) return 0; else return 2;
+    case 1: if (rfFreq == 6) return 1; else return 4;
+    case 2: if (rfFreq == 6) return 3; else return 5;
+    case 3: if (rfFreq == 6) return 5; else return 6;
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:76]: Lint Error:
```
73:  Else clause should never be on same line as else (use 2 lines)  
73:  If/else bodies with multiple statements require braces  
74:  Else clause should never be on same line as else (use 2 lines)  
74:  If/else bodies with multiple statements require braces  
75:  Else clause should never be on same line as else (use 2 lines)  
75:  If/else bodies with multiple statements require braces  
76:  Else clause should never be on same line as else (use 2 lines)  
76:  If/else bodies with multiple statements require braces  
```

File: src/protocol/crsf_uart.c [hunk 2]
```diff
@@ -29,6 +29,95 @@
...
+}
+
+static u8 convertElrsToPktRate(u8 rfFreq, u8 elrsRate)
+{
+  switch (elrsRate) {
+    case 0: return 0; //2.4
+    case 1: return 1; //2.4
+    case 2: return 0; //900
+    case 3: return 2; //2.4
+    case 4: return 1; //900
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:89]: Lint Error:
```
85:  At least two spaces is best between code and comments  
85:  Should have a space between // and comment  
86:  At least two spaces is best between code and comments  
86:  Should have a space between // and comment  
87:  At least two spaces is best between code and comments  
87:  Should have a space between // and comment  
88:  At least two spaces is best between code and comments  
88:  Should have a space between // and comment  
89:  At least two spaces is best between code and comments  
89:  Should have a space between // and comment  
```

File: src/protocol/crsf_uart.c [hunk 3]
```diff
@@ -29,6 +29,95 @@
...
+    case 5: if (rfFreq == 6) return 3; else return 2;
+    case 6: if (rfFreq == 6) return 4; else return 3;
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:91]: Lint Error:
```
90:  Else clause should never be on same line as else (use 2 lines)  
90:  If/else bodies with multiple statements require braces  
91:  Else clause should never be on same line as else (use 2 lines)  
91:  If/else bodies with multiple statements require braces  
```

File: src/protocol/crsf_uart.c [hunk 4]
```diff
@@ -29,6 +29,95 @@
...
+}
+
+static u16 convertPktRateToPeriod(u8 rfFreq, u8 rate)
+{
+  if (rfFreq == 0) return CRSF_FRAME_PERIOD;
+  switch (rate) {
+    case 0: if (rfFreq == 6) return 2000; else return 5000;
+    case 1: if (rfFreq == 6) return 4000; else return 10000;
+    case 2: if (rfFreq == 6) return 6666; else return 20000;
+    case 3: if (rfFreq == 6) return 20000; else return 40000;
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:103]: Lint Error:
```
100:  Else clause should never be on same line as else (use 2 lines)  
100:  If/else bodies with multiple statements require braces  
101:  Else clause should never be on same line as else (use 2 lines)  
101:  If/else bodies with multiple statements require braces  
102:  Else clause should never be on same line as else (use 2 lines)  
102:  If/else bodies with multiple statements require braces  
103:  Else clause should never be on same line as else (use 2 lines)  
103:  If/else bodies with multiple statements require braces  
```

File: src/protocol/crsf_uart.c [hunk 5]
```diff
@@ -163,6 +252,25 @@ static void processCrossfireTelemetryFrame()
   }
 }
 
+static void processElrsParamsData()
+{
+  if (telemetryRxBuffer[2] == TYPE_SETTINGS_WRITE && telemetryRxBuffer[3] == ADDR_RADIO && telemetryRxBuffer[4] == ADDR_MODULE && telemetryRxBuffer[5] == 0xFF) {
+    if (telemetryRxBuffer[6] == 0x01) currentBind = 1; else currentBind = 0;
+    if (telemetryRxBuffer[6] == 0x02) currentWiFi = 1; else currentWiFi = 0;
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:259]: Lint Error:
```
258:  Else clause should never be on same line as else (use 2 lines)  
258:  If/else bodies with multiple statements require braces  
259:  Else clause should never be on same line as else (use 2 lines)  
259:  If/else bodies with multiple statements require braces  
```

File: src/protocol/crsf_uart.c [hunk 6]
```diff
@@ -29,6 +29,115 @@
...
+  NULL
+};
+
+static u8 convertPktRateToElrs(u8 rfFreq, u8 rate)
+{
+  switch (rate) {
+    case 0: 
+      if (rfFreq == 6) return 0;
+      return 2;
+    case 1:
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:76]: Lint Error:
```
73:  Line ends in whitespace.  Consider deleting these extra spaces.  
76:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/crsf_uart.c [hunk 7]
```diff
@@ -29,6 +29,115 @@
...
+    case 0: 
+      if (rfFreq == 6) return 0;
+      return 2;
+    case 1: 
+      if (rfFreq == 6) return 1;
+      return 4;
+    case 2: 
+      if (rfFreq == 6) return 3;
+      return 5;
+    case 3:
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:82]: Lint Error:
```
79:  Line ends in whitespace.  Consider deleting these extra spaces.  
82:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/crsf_uart.c [hunk 8]
```diff
@@ -29,6 +29,115 @@
...
+  switch (elrsRate) {
+    case 0: return 0;   // 2.4
+    case 1: return 1;   // 2.4
+    case 2: return 0;   // 900
+    case 3: return 2;   // 2.4
+    case 4: return 1;   // 900
+    case 5: 
+      if (rfFreq == 6) return 3;
+      return 2;
+    case 6:
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:101]: Lint Error:
```
98:  Line ends in whitespace.  Consider deleting these extra spaces.  
101:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/crsf_uart.c [hunk 9]
```diff
@@ -29,6 +29,115 @@
...
+      return 3;
+  }
+  return 0;
+}
+
+static u16 convertPktRateToPeriod(u8 rfFreq, u8 rate)
+{
+  if (rfFreq == 0) return CRSF_FRAME_PERIOD;
+  switch (rate) {
+    case 0:
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:112]: Lint Error:
```
112:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/crsf_uart.c [hunk 10]
```diff
@@ -29,6 +29,115 @@
...
+      if (rfFreq == 6) return 2000;
+      return 5000;
+    case 1:
+      if (rfFreq == 6) return 4000;
+      return 10000;
+    case 2: 
+      if (rfFreq == 6) return 6666;
+      return 20000;
+    case 3: 
+      if (rfFreq == 6) return 20000;
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:122]: Lint Error:
```
118:  Line ends in whitespace.  Consider deleting these extra spaces.  
121:  Line ends in whitespace.  Consider deleting these extra spaces.  
122:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/crsf_uart.c [hunk 11]
```diff
@@ -163,6 +272,35 @@ static void processCrossfireTelemetryFrame()
   }
 }
 
+static void processElrsParamsData()
+{
+  if (telemetryRxBuffer[2] == TYPE_SETTINGS_WRITE && telemetryRxBuffer[3] == ADDR_RADIO && telemetryRxBuffer[4] == ADDR_MODULE && telemetryRxBuffer[5] == 0xFF) {
+    if (telemetryRxBuffer[6] == 0x01) {
+      currentBind = 1;
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:279]: Lint Error:
```
279:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/crsf_uart.c [hunk 12]
```diff
@@ -163,6 +272,35 @@ static void processCrossfireTelemetryFrame()
...
+    } else {
+      currentBind = 0;
+    }
+
+    if (telemetryRxBuffer[6] == 0x02) {
+      currentWiFi = 1; 
+    } else {
+      currentWiFi = 0;
+    }
+
```
Comment (PhracturedBlue) [inline src/protocol/crsf_uart.c:289]: Lint Error:
```
285:  Line ends in whitespace.  Consider deleting these extra spaces.  
289:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

**Comment by Dropsleutel** [comment]
Hi Phobos, 
I flashed your firmware in my tx, just to have a look. 
just to mention I have no ESP 2.4ghz, only Frsky R9 I could test for you. 
R9M module works in via PXX, as protocol it is inverted in deviation. 
So CRSF output, it need to get inversion option in order to use inverted R9M in the back.

I see that firmware, nr c84450e it is excact same nr, as version i flashed in my Tx 30december2020. 

There is no ELRS protocol in the list, but when selecting CRSF  protocol, (note no module present in bay), I can see an empty menu, no txt, just yellow line which can be changed 4 positions up/down.

Thanks for all your great help and efforts !!
let me know if can help testing

**Comment by phobos-** [comment]
Hi, thanks for the comment.
Hexfet went to hell and back to invert the PXX protocol :) 
He basically implemented it as a bitbanged PWM:
https://github.com/DeviationTX/deviation/blob/master/src/target/tx/devo/common/protocol/pwm.c#L118

For CRSF it is different, it is a hardware UART implementation:
https://github.com/DeviationTX/deviation/blob/master/src/protocol/crsf_uart.c#L286

Deviation uses STM32F103RET6 at least for t8sg which I have on hand - that processor is incapable of inverting the UART signal in software, there are additional hardware inverters needed. That would not be the case if an F3 family was used.
What could be done from the software side is implementing a software serial communication similar to PXX, but then I don't think we would be able to achieve 400k baudrate speed required by CRSF protocol.

I actually also have an R9M module which I flashed and modded to work with ExpressLRS and Deviation, but I modded it in a way that only allows for one way communication between the module and TX. What it means is you can only send stick information to the module, but you will not receive any communication back, like configuration data, telemetry, etc. If you still want to do the mod you can refer to the short tutorial I posted on RCGroups:
https://www.rcgroups.com/forums/showthread.php?3437865-ExpressLRS-DIY-LoRa-based-race-optimized-RC-link-system/page13#post46673237

**Comment by dat-alex** [comment]
I have recompiled source for the 2400_TX_ESP32_SX1280_E28_via_UART **BUT I have not eps32 TX module available to test** do you want to upload here for someone to test?

[DIY_2400_TX_ESP32_SX1280_E28_via_UART.zip](https://github.com/DeviationTX/deviation/files/6324382/DIY_2400_TX_ESP32_SX1280_E28_via_UART.zip)

(this way you have not to do the 'dirty woprk' of installing Visual Studio, settings ,for compiling etc....)

**Comment by Dropsleutel** [comment]
@Phobos 

And what about the R9M, i assume there is an hardware inverter onboard?  
Maybe bridging or bypassing it ? Direct feed to the uart pin at R9M cpu?

The resistor mod and alos the temperature issues in R9M were serial comms are affected, maybe we could also get rid off as bot bit of strange issues....

Is there any shematic of R9M or the Happymodel 868?

**Comment by tonvfx** [comment]
Hey! 

This is great. Would it be possible to also to a build for the V2 non-plus? The Carbon T8SG uses that firmware. Thanks!

**Comment by phobos-** [comment]
@tonvfx 

[deviation-t8sg_v2-v5.0.0-575ef03.zip](https://github.com/DeviationTX/deviation/files/6398714/deviation-t8sg_v2-v5.0.0-575ef03.zip)

**Comment by danmeam** [comment]
with diy esp-01f+sx127x 868mhz it works fine

https://github.com/ExpressLRS/ExpressLRS/wiki/Building-an-Esp-Rx#custom-esp-900-mhz-rx

**Comment by tonvfx** [comment]
> @tonvfx
> 
> [deviation-t8sg_v2-v5.0.0-575ef03.zip](https://github.com/DeviationTX/deviation/files/6398714/deviation-t8sg_v2-v5.0.0-575ef03.zip)

That's awesome!! Thanks.. As soon as my module arrives I'll get going with this. Thank you very much!

**Comment by tonvfx** [comment]
> @tonvfx
> 
> [deviation-t8sg_v2-v5.0.0-575ef03.zip](https://github.com/DeviationTX/deviation/files/6398714/deviation-t8sg_v2-v5.0.0-575ef03.zip)

Got it up and running. Nice one.

**Comment by danmeam** [comment]
Thats great, have you tried on fligth allready, mine fell a loot more reactive...

**Comment by phobos-** [comment]
I pushed a small bugfix that caused the menus to glitch sometimes. Builds are updated in the first (top) comment.
If anyone is interested in a build for a different target - shout here, I'll upload it 😅

**Comment by jviticofpv** [comment]
Hello my friends... 

I would like a little help over here, just to be clear that my **Jumper/RadioMaster TX8 (deviation-t8sg_v2)** would work ok with Happymodel 2.4Ghz TX/RX combo.

**- In order to make it work I must do the next steps??**

**1- Crossfire pin mod.
2- Flash this firmware ((deviation-t8sg_v2-v5.0.0-1dc62dc.zip))
3- Update Tx and Rx via WIFI with my pc
4- and that's it??**

I would appreciate your help.... not understanding what to do... it's what have me with my purchase on hold.


Thanks in advance from Chile (Venezuelan guy)

**Comment by Thomas-Lawford** [comment]
Hi, I installed the firmware and completed the hardware mod, but I was wondering:
do I need to add a LUA script? -if yes where in the files am I supposed to add it, and where can  I get this script?

**Comment by phobos-** [comment]
@jviticofpv yes, that's it, just remember to flash Happymodel tx with -DUART_INVERTED commented out :
```#-DUART_INVERTED```

**Comment by phobos-** [comment]
@Thomas-Lawford no need, just follow the same steps as @jviticofpv described in the comment above yours.

**Comment by jviticofpv** [comment]
> @jviticofpv yes, that's it, just remember to flash Happymodel tx with -DUART_INVERTED commented out :
> `#-DUART_INVERTED`

perfect.... but this part is what I don't understand.   I'm sorry to insist, but could you tell me where to get the "how to" about this part.... specially the "comment out part"...

if this means that I must do some coding? (installing coding software)?
or just get the happymodel firmware, look for "USER_DEFINES.TXT" file, and comment out with quotes the line that you mention?

if is like this last part (I'll hope) will. be perfect, because I know that I can do it 😅😅..


thanks again for your help...

**Comment by phobos-** [comment]
@jviticofpv You can use the ExpressLRS configurator to flash the firmware:
https://www.youtube.com/watch?v=XpCvYpyQ4LM&t=146s&ab_channel=KaijuKaiju
https://www.youtube.com/watch?v=wBUamLm51XQ&t=13s&ab_channel=UAVTechUAVTech

here are two examples, there are many more on youtube.

**Comment by jviticofpv** [comment]
> @jviticofpv You can use the ExpressLRS configurator to flash the firmware:
> https://www.youtube.com/watch?v=XpCvYpyQ4LM&ab_channel=KaijuKaiju
> https://www.youtube.com/watch?v=wBUamLm51XQ&ab_channel=UAVTechUAVTech
> 
> here are two examples, there are many more tutorials on youtube.
> Finally, if you'll still have trouble with flashing expresslrs stuff you can hop on discord, there are many knowledgable people that can help with specific issues:
> https://discord.com/invite/dS6ReFY

thanks so much.... I'm already purchased my expresslrs setup... thanks for clarify my doubts....

**Comment by tonvfx** [comment]
> I pushed a small bugfix that caused the menus to glitch sometimes. Builds are updated in the first (top) comment.
> If anyone is interested in a build for a different target - shout here, I'll upload it 😅

Hey! I just flashed this to my radio, seemed fine.. except, when I accidentally switched to sbus and back to crsf the menu was gone. I had to load another model, switch off the radio, switch on, reload elrs and it was back. Also, I doesn't look like wifi update is working on MacOs.. Chrome stalls at 17% and then wifi is lost... which brings me to another question, is there a way for me to see what version of expresslrs I'm running on my radio? ("version" in the menu would be nice..)

Thanks for the awesome work!

**Comment by howels** [comment]
> > I pushed a small bugfix that caused the menus to glitch sometimes. Builds are updated in the first (top) comment.
> > If anyone is interested in a build for a different target - shout here, I'll upload it 😅
> 
> Hey! I just flashed this to my radio, seemed fine.. except, when I accidentally switched to sbus and back to crsf the menu was gone. I had to load another model, switch off the radio, switch on, reload elrs and it was back. Also, I doesn't look like wifi update is working on MacOs.. Chrome stalls at 17% and then wifi is lost... which brings me to another question, is there a way for me to see what version of expresslrs I'm running on my radio? ("version" in the menu would be nice..)
> 
> Thanks for the awesome work!

Wait for eLRS 1.1, it'll arrive soon and will work using the same code as existing CRSF support in deviation nightly builds. No need for the special build above (which is probably causing the issue you report with menus). No idea why WiFi update doesn't work- please report on eLRS GitHub.

**Comment by tonvfx** [comment]
> > > I pushed a small bugfix that caused the menus to glitch sometimes. Builds are updated in the first (top) comment.
> > > If anyone is interested in a build for a different target - shout here, I'll upload it 😅
> > 
> > 
> > Hey! I just flashed this to my radio, seemed fine.. except, when I accidentally switched to sbus and back to crsf the menu was gone. I had to load another model, switch off the radio, switch on, reload elrs and it was back. Also, I doesn't look like wifi update is working on MacOs.. Chrome stalls at 17% and then wifi is lost... which brings me to another question, is there a way for me to see what version of expresslrs I'm running on my radio? ("version" in the menu would be nice..)
> > Thanks for the awesome work!
> 
> Wait for eLRS 1.1, it'll arrive soon and will work using the same code as existing CRSF support in deviation nightly builds. No need for the special build above (which is probably causing the issue you report with menus). No idea why WiFi update doesn't work- please report on eLRS GitHub.

Wait? Haha.. with something this exciting? ;-) .. But that's really good to hear. I'll ask about the wifi update 'issue' on the elrs github/ discord... Like I said.. it's also possible it does work, but there's (in deviationtx at least atm) not an easy way to check version running on the tx afaik....

**Comment by danmeam** [comment]
> > > > I pushed a small bugfix that caused the menus to glitch sometimes. Builds are updated in the first (top) comment.
> > > > If anyone is interested in a build for a different target - shout here, I'll upload it 😅
> > > 
> > > 
> > > Hey! I just flashed this to my radio, seemed fine.. except, when I accidentally switched to sbus and back to crsf the menu was gone. I had to load another model, switch off the radio, switch on, reload elrs and it was back. Also, I doesn't look like wifi update is working on MacOs.. Chrome stalls at 17% and then wifi is lost... which brings me to another question, is there a way for me to see what version of expresslrs I'm running on my radio? ("version" in the menu would be nice..)
> > > Thanks for the awesome work!
> > 
> > 
> > Wait for eLRS 1.1, it'll arrive soon and will work using the same code as existing CRSF support in deviation nightly builds. No need for the special build above (which is probably causing the issue you report with menus). No idea why WiFi update doesn't work- please report on eLRS GitHub.
> 
> Wait? Haha.. with something this exciting? ;-) .. But that's really good to hear. I'll ask about the wifi update 'issue' on the elrs github/ discord... Like I said.. it's also possible it does work, but there's (in deviationtx at least atm) not an easy way to check version running on the tx afaik....

Well.. i think if update dont work and it work on rx, they cant bind if both dont have same version..

**Comment by Dropsleutel** [comment]
Just to inform I have r9m running in t8sg v2 plus fine. (inc telemetry)
I have flashed r9m with ELRS and it runs in openTx radio.
Now I connected as such;

r9m_backpack_pins

I connected Rx1, to the CPPM pin. (and removed the transistor)
I connected Tx Alternative location to Smartport pin. 

I flashed https://github.com/DeviationTX/deviation/files/6503349/deviation-t8sg_v2_plus-v5.0.0-1dc62dc.zip

However in CRSF menu,  I get an empty bar which I can move up and down.
I don't see any Txt or data in the menus.

**Comment by howels** [comment]
> Just to inform I had r9m running in t8sg v2 plus fine. (inc telemetry) (modified the tray)
> I flashed my r9m with ELRS firmware and it runs in OpenTx radio.
> Now I connected as such;
> 
> r9m_backpack_pins
> 
> I connected Rx1, to the CPPM pin. (and removed the transistor)
> I connected Tx Alternative location to Smartport pin.
> 
> I flashed https://github.com/DeviationTX/deviation/files/6503349/deviation-t8sg_v2_plus-v5.0.0-1dc62dc.zip
> 
> However in CRSF menu, I get an empty bar which I can move up and down.
> I don't see any Txt or data in the menus.

Did you change the UART inversion when building eLRS? Setting for T8SG is opposite of OpenTX.

**Comment by Dropsleutel** [comment]
I used R9M.   Inversion is only availble for DIY ESP?  

However I connected direct to tx rx pins as by picture above, so it should bypass the Frsky invertors on the pcb of the R9M ?

**Comment by Teiho69** [comment]
Just flashed my T8SG v2 carbon with deviation-t8sg_v2-v5.0.0-575ef03 .
My Happymodel ES900 tx works like a charm, so do my TBS Tracer tx.
I'm glad that CSRF support for TBS Tx is not lost.
Thank you really much.

**Comment by kotyatokino** [comment]
> @kotyatokino this change is what BREAKS it, not fixes it. I removed this line from expresslrs code, compiled & flashed my ExpressLRS TX module and it started working again.
> 
> I changed: `CRSF::duplex_set_TX();` -> `//CRSF::duplex_set_TX();`

Thank you for your explation.
CRSF::duplex_set_TX() will break the way to CRSF_Cmds().
We need to find reason why . I will try it with my poor understanding.

**Comment by kotyatokino** [comment]
hmmm.. I just giving up debugging...
When we see debug print , I need to buy phono plug UART converter cable..

**Comment by howels** [comment]
> hmmm.. I just giving up debugging... When we see debug print , I need to buy phono plug UART converter cable..

Yes, it's the most awkward debug system. I thought the same thing.

**Comment by Dropsleutel** [comment]
I just gave it a re-try T8SG V2 Plus, with R9M ExpressLRS flashed, but the CRSF menu is not working for me.

**Comment by dat-alex** [comment]
Hi ! I 've just received my new HappyModel elrs 2.4 Tx module and trying it with T8SG plus .I get the classic problem with blankCRSF menus.
Is there any Deviation fw build compatible with ELRS F4 2G4 AIO Flight Controller spi receiver already compiled for the plus with OLED (like Phobos no-OLED)?

Thank you for your time
Alex

**Comment by howels** [comment]
> Hi ! I 've just received my new HappyModel elrs 2.4 Tx module and trying it with T8SG plus .I get the classic problem with blankCRSF menus. Is there any Deviation fw build compatible with ELRS F4 2G4 AIO Flight Controller spi receiver already compiled for the plus with OLED (like Phobos no-OLED)?
> 
> Thank you for your time Alex

Try this https://github.com/DeviationTX/deviation/pull/1018#issue-1052703851

**Comment by harlock00999** [comment]
Hi, First things first, thank you all for trying to keep Deviationtx alive and usable with ELRS! It is thanks to this post if I'm able to use a HGLRC TX 2.4Ghz on my beloved T8SG V2 PLUS! - THANK YOU ALL - (@phobos- you're the best)

Now, as many of you I was stuck with fw 1.1 since I read that 1.2 wasn't good yet.

But yesterday I flashed my tx with ELSR fw version 1.2.1 and everything seem to work just fine. The only bug I noticed is the power value that you can add on the main page of your model freeze sometimes and come back if you go in the model setup and re-int the CRFS module. Other values like LQ, RSSI work fine on both radio and Betaflight.  

Thoughts? Questions? Comments? Do you want me to do some tests? 
Please keep in mind that I'm just a user...well, a smart user if you will ^_^

Thank you guys and happy flying

**Comment by StonedDawg** [comment]
@harlock00999 elrs stable V2 has been released, please use V2, and this PR from Phobos  for your deviationtx 
[Phobos deviationtx for elrs V2](https://github.com/DeviationTX/deviation/pull/1018#issue-1052703851)

**Comment by harlock00999** [comment]
> @harlock00999 elrs stable V2 has been released, please use V2, and this PR from Phobos for your deviationtx
> [Phobos deviationtx for elrs V2](https://github.com/DeviationTX/deviation/pull/1018#issue-1052703851)

I'm going right now!!! Thank you!

**Comment by phobos-** [comment]
closing this, as expresslrs v2.0 is natively supported in deviationTX.

**Comment by howels** [comment]
> closing this, as expresslrs v2.0 is natively supported in deviationTX.

There's still the 12 character limit, which many users struggle with. I'm using -DDEVICE_NAME to compile my eLRS TX with a shorter name, but would be great if the CRSF code could just concat fields to 12 chars instead of rendering a blank line.

**Pull request closed without merging**